### ISSUE TITLE
PDF/A PDF/UA Support

### ISSUE THREAD

**Issue by pipebaum** [LibrePDF/OpenPDF#493]
Trying to figure out if either PDF/A or PDF/UA are supported. If so what versions and is there any documentation?

**Comment by asturio** [comment]
Yes, PDF/A is supported. Take a look in the Class `PdfWriter`. You can use the method `setPDFXConformance()` and some of the Constants there, like `PDFA1A` or `PDFA1B`.
There is no example for that, but you can create one in pdf-toolbox if you want.
Pull requests are welcome.

**Comment by andreasrosdal** [comment]
What is needed to support PDF/UA?
https://en.m.wikipedia.org/wiki/PDF/UA

**Comment by arnthom** [comment]
In Germany the "Barrierefreiheitsstärkungsgesetz" (accessibility strenghtening act) comes into force in June 2025. 
All companies (like insurance etc) have to ensure, that their pdf documents are compatible and accessible i.e. for blind people.
Therefore I look for the parameters to set to ensure pdf/a compatibility.
There is a check tool available at https://pdfua.foundation/de/ 

When doing a short check for my pdf created with OpenPdf 2.0.2 i see
- no title
- no xmp meta data
- no language defined for the document
- no language defines for text elements 
- no embedded fonts

More Information available at
- https://www.bmas.de/DE/Service/Gesetze-und-Gesetzesvorhaben/barrierefreiheitsstaerkungsgesetz.html
- https://pdfa.org/resource/the-matterhorn-protocol/

Any news on pdf/ua?

**Comment by mkl-public** [comment]
Most of the items you list are things you can already do easily with OpenPdf, you merely are not forced to do them by the software, you have to remember doing them yourself.

**Comment by arnthom** [comment]
I use openpdf as a library creating the pdf documentation for my software product.
I cannot patch every pdf that is created, i have to do it with the openpdf software.
So there should be a way to define the xmp meta data or the document's language with methods of pdfwriter or pdfdocument.

But maybe ou can tell me how to do it "easily with OpenPdf", perhaps I'll find a wworkaround.

**Comment by mkl-public** [comment]
> So there should be a way to define the xmp meta data or the document's language with methods of pdfwriter or pdfdocument.

You might want to look at `PdfWriter.createXmpMetadata` that creates metadata from existing info data (or `PdfWriter.setXmpMetadata` if you want to create the metadata yourself, e.g. with Adobe's [XMP Toolkit for Java](https://www.adobe.com/devnet/xmp.html)).

The document language is but an entry in the PDF catalog dictionary.

I would assume, though, that creating PDF/UA conform tagging will be the largest challenge.

**Comment by arnthom** [comment]
Neither of the document.add-methods for Subject, Keywords, Author and Creator nor document,setDocumentLanguage() make their way to the Pdf catalog dictionary.
writer.createXmpMetadata() does not seem to make a difference, although the methods mentioned above have been called.

Would it be possible as a first step to write this catalog dictionary into the pdf file?

**Comment by arnthom** [comment]
any help would be appreciated.

**Comment by shashankbenakatti** [comment]
Hey , is the support added for this ? I would like to work on this. From what I can see the issue is still open.

**Comment by asturio** [comment]
@shashankbenakatti sure, Any help is welcome.

**Comment by arnthom** [comment]
with a little help of ... AI I found how to define the pdf language
`writer.getExtraCatalog().put(PdfName.LANG, new PdfString("de-DE"));`

with 
`writer.getExtraCatalog().put(PdfName.MARKINFO, new PdfDictionary(PdfName.MARKED));`
the pdf/ua compatibility check does not complain about missing structure.

but the IAccessibleElement interface seems to be missing.
Is there a way to set things like image description?

**Comment by arnthom** [comment]
finally i managed to fix many of the issues pdf accessibility checker showed.

1. use `writer.setEncryption(null, null, PdfWriter.ALLOW_SCREENREADERS, PdfWriter.STANDARD_ENCRYPTION_128);`
2. set language with `writer.getExtraCatalog().put(PdfName.LANG, new PdfString("de-DE"));` 
3. use '		PdfDictionary markInfo = new PdfDictionary();
		markInfo.put(PdfName.MARKED, PdfBoolean.PDFTRUE);
		writer.getExtraCatalog().put(PdfName.MARKINFO, markInfo);' 
4. real tagging is not possible, when using document.add() 
5. set the document title '		writer.setViewerPreferences(PdfWriter.DisplayDocTitle);'
6. create xmp metadata
'		// --- XMP manuell erzeugen (korrektes dc:title als LangAlt) ---
		OutputStream baos = new ByteArrayOutputStream();
		try {
			XmpWriter xmp = new XmpWriter(baos);
			DublinCoreSchema dc = new DublinCoreSchema();
			LangAlt title = new LangAlt();
			title.addLanguage("x-default", pdftitle);
			dc.setProperty(DublinCoreSchema.TITLE, title);
			xmp.addRdfDescription(dc);

		    // --- PDF/UA Identifier 
		    XmpSchema pdfUaId = new XmpSchema("xmlns:pdfuaid=\"http://www.aiim.org/pdfua/ns/id/\"") {
		        // anonyme Subklasse – absichtlich leer
		    };
		    pdfUaId.setProperty("pdfuaid:part", "1"); // Pflicht für PDF/UA-1
		    xmp.addRdfDescription(pdfUaId);

			xmp.close();
			writer.setXmpMetadata(((ByteArrayOutputStream) baos).toByteArray());
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
			writer.createXmpMetadata();
		}
'

**Comment by asturio** [comment]
Maybe having a better API/Interface for doing this more easily would be nice. Maybe having a data structure for all the input you need, like:
- language
- encryption(s)
- title
- and so on...

Than calling a new method like .makeDocumentAccessible (please create a better name :-) )

Can you imagine such an interface?

**Comment by arnthom** [comment]
such an API would be nice.
besides it would be nice to have support for structured PDF in the high-level api, where i don't have to worry about placing the elements ...
